components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/internal/core/ApplicationAuthenticatorManager.java

```diff
@@ -0,0 +1,159 @@
+     */
+    public ApplicationAuthenticator getSystemDefinedAuthenticatorByName(String authenticatorName) {
+
+        for (ApplicationAuthenticator authenticator : systemDefinedAuthenticators) {
+            if (authenticator.getName().equals(authenticatorName)) {
+                return authenticator;
+            }
+        }
+        return null;
+    }
+
+    /**
+     * Get all the authenticators for the given tenant domain.
+     *
+     * @param tenantDomain Tenant domain.
+     * @return List of authenticators.
+     */
+    public List<ApplicationAuthenticator> getAllAuthenticators(String tenantDomain) {
+
+        List<ApplicationAuthenticator> allAuthenticators = new ArrayList<>(systemDefinedAuthenticators);
+
+        try {
+            for (LocalAuthenticatorConfig localConfig : ApplicationAuthenticatorService.getInstance()
+                    .getAllUserDefinedLocalAuthenticators(tenantDomain)) {
+                allAuthenticators.add(FrameworkServiceDataHolder.getInstance().getAuthenticatorAdapterService()
+                        .getLocalAuthenticatorAdapter(localConfig));
+            }
+
+            FederatedAuthenticatorConfig[] fedConfig = IdentityProviderManager.getInstance()
+                    .getAllFederatedAuthenticators(tenantDomain);
+            for (FederatedAuthenticatorConfig fedAuth : fedConfig) {
+                allAuthenticators.add(FrameworkServiceDataHolder.getInstance().getAuthenticatorAdapterService()
+                        .getFederatedAuthenticatorAdapter(fedAuth));
+            }
+
+            return allAuthenticators;
+        } catch (Exception e) {
+            throw new RuntimeException("Error while getting all application authenticators.", e);
+        }
+    }
+
+    /**
+     * Get the ApplicationAuthenticator for the given authenticator name.
+     *
+     * @param authenticatorName Authenticator name.
+     * @param tenantDomain      Tenant domain.
+     * @return  ApplicationAuthenticator instance.
+     */
+    public ApplicationAuthenticator getApplicationAuthenticatorByName(String authenticatorName, String tenantDomain) {
+
+        // Check whether the authenticator is in the system defined authenticator.
+        for (ApplicationAuthenticator authenticator : systemDefinedAuthenticators) {
+            if (authenticator.getName().equals(authenticatorName)) {
+                return authenticator;
+            }
+        }
+
+        // Check whether the authenticator config is the user defined local authenticator config, if so resolve it.
+        try {
+            LocalAuthenticatorConfig localConfig = ApplicationAuthenticatorService.getInstance()
+                    .getUserDefinedLocalAuthenticator(tenantDomain, authenticatorName);
+            if (localConfig != null) {
+                return FrameworkServiceDataHolder.getInstance().getAuthenticatorAdapterService()
+                        .getLocalAuthenticatorAdapter(localConfig);
+            }
```

Comment:
ApplicationAuthenticatorService.getInstance()
                    .getUserDefinedLocalAuthenticator(tenantDomain, authenticatorName) -> ApplicationAuthenticatorService.getInstance()
                    .getUserDefinedLocalAuthenticatorConfig(tenantDomain, authenticatorName)
 
FrameworkServiceDataHolder.getInstance().getAuthenticatorAdapterService()
                        .getLocalAuthenticatorAdapter(localConfig) -> FrameworkServiceDataHolder.getInstance().getUserDefinedAuthenticatorService()
                        .getUserDefinedLocalAuthenticator(localConfig)

Reply:
Not sure to which level above change is feasible. Let's discuss based on feasibility

Reply:
Will fix this.